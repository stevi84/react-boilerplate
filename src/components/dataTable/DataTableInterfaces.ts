import { ColDef } from 'ag-grid-community';

export enum CrudMode {
  CREATE = 'CREATE',
  READ = 'READ',
  UPDATE = 'UPDATE',
}

export interface EntityManager<EntityType> {
  create: () => void;
  read: () => void;
  update: (entity: EntityType) => void;
  delete: (entity: EntityType) => void;
  getEmpty: () => EntityType;
}

export interface DataTableOwnProps<EntityType> {
  id: string;
  columns: ColDef<EntityType>[];
  rowsData: EntityType[];
  manager: EntityManager<EntityType>;
}
