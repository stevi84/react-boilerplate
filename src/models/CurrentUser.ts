import { Components } from '../apis/api';

export type CurrentUser = Components.Schemas.CurrentUser;
export type Role = Components.Schemas.Role;
export const getEmptyCurrentUser = (): CurrentUser => ({
  loginName: '',
  firstName: '',
  surName: '',
  mail: '',
  roles: [],
});
