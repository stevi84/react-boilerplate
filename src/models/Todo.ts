import { Components } from '../apis/api';

export type Todo = Components.Schemas.Todo;

export const getEmptyTodo = (): Todo => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  owner: '',
  dueDate: '',
  description: '',
  completed: false,
});
