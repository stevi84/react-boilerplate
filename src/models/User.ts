import { Components } from '../apis/api';

export type User = Components.Schemas.User;

export const getEmptyUser = (): User => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  name: '',
  dateOfBirth: '',
  size: 0,
  weight: 0,
  email: '',
  phone: '',
});
