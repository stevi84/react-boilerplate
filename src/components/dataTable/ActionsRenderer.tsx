import { IconButton, Tooltip } from '@mui/material';
import EditIcon from '@mui/icons-material/Edit';
import DeleteIcon from '@mui/icons-material/Delete';
import { useTranslation } from 'react-i18next';
import { BaseEntity } from '../../models/BaseEntity';
import { EntityManager } from './DataTableInterfaces';
import { ICellRendererParams } from 'ag-grid-community';
import { Authorization } from '../common/Authorization';

interface ButtonCellRendererProps<EntityType> {
  id: string;
  manager: EntityManager<EntityType>;
}

export const ActionsRenderer = <EntityType extends BaseEntity>(
  props: ButtonCellRendererProps<EntityType> & ICellRendererParams<EntityType>
) => {
  const { id, data, manager } = props;

  const { t } = useTranslation();

  return (
    <Authorization
      allowedAccessRights={[manager.editRight]}
      WrappedElement={
        <>
          <Tooltip title={t('edit')}>
            <span>
              <IconButton id={`${id}-${data!.id}-button-edit`} color="primary" onClick={() => manager.update(data!)}>
                <EditIcon />
              </IconButton>
            </span>
          </Tooltip>
          <Tooltip title={t('delete')}>
            <span>
              <IconButton id={`${id}-${data!.id}-button-delete`} color="primary" onClick={() => manager.delete(data!)}>
                <DeleteIcon />
              </IconButton>
            </span>
          </Tooltip>
        </>
      }
    />
  );
};
