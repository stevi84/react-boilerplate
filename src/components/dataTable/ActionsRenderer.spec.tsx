import { describe, it, expect, vi } from 'vitest';
import { createRenderer, ShallowRenderer } from 'react-test-renderer/shallow';
import { fireEvent, render, screen } from '@testing-library/react';
import { ActionsRenderer } from './ActionsRenderer';
import { EntityManager } from './DataTableInterfaces';
import { BaseEntity } from '../../models/BaseEntity';

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

const updateMock = vi.fn();
const deleteMock = vi.fn();
const managerMock = { update: updateMock, delete: deleteMock } as unknown as EntityManager<BaseEntity>;

describe('ActionsRenderer', () => {
  it('should equal saved snapshot', () => {
    const renderer: ShallowRenderer = createRenderer();
    // @ts-ignore
    renderer.render(<ActionsRenderer id="id" manager={managerMock} data={{ id: 1 } as any} />);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('should trigger update on button click', () => {
    // @ts-ignore
    render(<ActionsRenderer id="id" manager={managerMock} data={{ id: 1 } as any} />);
    expect(updateMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('EditIcon'));
    expect(updateMock.mock.calls.length).toEqual(1);
  });

  it('should trigger delete on button click', () => {
    // @ts-ignore
    render(<ActionsRenderer id="id" manager={managerMock} data={{ id: 1 } as any} />);
    expect(deleteMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('DeleteIcon'));
    expect(deleteMock.mock.calls.length).toEqual(1);
  });
});
