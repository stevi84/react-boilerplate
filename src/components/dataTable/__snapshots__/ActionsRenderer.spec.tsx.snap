// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionsRenderer > should equal saved snapshot 1`] = `
<DocumentFragment>
  <span
    aria-label="edit"
    class=""
    data-mui-internal-clone-element="true"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
      id="id-1-button-edit"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="EditIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </span>
  <span
    aria-label="delete"
    class=""
    data-mui-internal-clone-element="true"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
      id="id-1-button-delete"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </span>
</DocumentFragment>
`;
